src/Functions/seriesPeriodDetectFFT.cpp

```diff
@@ -0,0 +1,205 @@
+ * 3. Find the peak value (highest) for dominant frequency component.
+ * 4. Inverse of the dominant frequency component is the period.
+*/
+
+class FunctionSeriesPeriodDetectFFT : public IFunction
+{
+public:
+    static constexpr auto name = "seriesPeriodDetectFFT";
+
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionSeriesPeriodDetectFFT>(); }
+
+    std::string getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors args{{"time_series", &isArray<IDataType>, nullptr, "Array"}};
+        validateFunctionArgumentTypes(*this, arguments, args);
+
+        return std::make_shared<DataTypeFloat64>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        ColumnPtr array_ptr = arguments[0].column;
+        const ColumnArray * array = checkAndGetColumn<ColumnArray>(array_ptr.get());
+
+        const IColumn & src_data = array->getData();
+
+        auto res = ColumnFloat64::create(input_rows_count);
+        auto & res_data = res->getData();
+
+        Float64 period;
+        for (size_t i = 0; i < input_rows_count; ++i)
+            if (executeNumber<UInt8>(src_data, period) || executeNumber<UInt16>(src_data, period) || executeNumber<UInt32>(src_data, period)
+                || executeNumber<UInt64>(src_data, period) || executeNumber<Int8>(src_data, period)
+                || executeNumber<Int16>(src_data, period) || executeNumber<Int32>(src_data, period)
+                || executeNumber<Int64>(src_data, period) || executeNumber<Float32>(src_data, period)
+                || executeNumber<Float64>(src_data, period))
+            {
+                res_data[i] = period;
+            }
+            else
+                throw Exception(
+                    ErrorCodes::ILLEGAL_COLUMN,
+                    "Illegal column {} of first argument of function {}",
+                    arguments[0].column->getName(),
+                    getName());
+
+        return res;
+    }
+
+    template <typename T>
+    bool executeNumber(const IColumn & src_data, Float64 & period) const
+    {
+        const ColumnVector<T> * src_data_concrete = checkAndGetColumn<ColumnVector<T>>(&src_data);
+        if (!src_data_concrete)
+            return false;
+
+        const PaddedPODArray<T> & src_vec = src_data_concrete->getData();
+
+        size_t len = src_vec.size();
+        if (len < 4)
+        {
+            period = NAN; // Atleast four data points are required to detect period
```

Comment:
Spellcheck.

Reply:
IMO the previous version with the `ErrorCodes::BAD_ARGUMENTS` exception was more explicit than returning 'nan'.

Reply:
yes, the exception was better in terms of explicit. But, it became imperative for me to change this to conform with CH design of returning exactly one value for every input row.